include/metal/detail/sfinae.hpp

```diff
@@ -58,6 +58,15 @@ namespace metal {
         template<template<typename...> class expr, typename... vals>
         using call = expr<vals...>;
 #endif
+
+        template<class... vals>
+        struct _dcaller {
+            template<template<class...> class expr>
+            using type = expr<vals...>;
+        };
+
+        template<template<class...> class expr, class... vals>
+        using dcall = typename _dcaller<vals...>::template type<expr>;
```

Comment:
I wonder if you can just use `metal::invoke` instead of `detail::dcall` with no impact to performance. Notice that `metal::invoke` is defined in terms of `metal::apply`, which at some point [used to be implemented exactly this way](https://github.com/brunocodutra/metal/blob/630e8feb1da4388ba5c08b1e5f432c58c24b429a/include/metal/lambda/apply.hpp#L70). I [later changed it](https://github.com/brunocodutra/metal/commit/2071e36f3bbcf6fbe57f94a568f34812ad581c7e#diff-5f827d47769fbe9312e9e904c05fbc3eR53) along with the introduction of `metal::cascade`, but I can't remember why, but likely due to some compiler that didn't want to cooperate. I don't know whether that change had any impact in benchmarks, but we could definitely attempt to revert it back to the previous implementation to avoid introducing `dcall`.

Reply:
> I wonder if you can just use `metal::invoke` instead of `detail::dcall` with no impact to performance. 

Thanks for the suggestion. It [does](https://ecrypa.github.io/metal-benchmark/min_element_clang9_v4_v15_bestof/) [work](https://ecrypa.github.io/metal-benchmark/min_element_gcc9_v4_v15_bestof/) (see [v15](https://github.com/brunocodutra/metal/commit/22844ee3eb780a8539a707a1fc088c839772795a)).

I had not expected that `invoke` is implemented in this way. I was assuming an alias.

> Notice that `metal::invoke` is defined in terms of `metal::apply`, which at some point [used to be implemented exactly this way](https://github.com/brunocodutra/metal/blob/630e8feb1da4388ba5c08b1e5f432c58c24b429a/include/metal/lambda/apply.hpp#L70). I later [changed it](https://github.com/brunocodutra/metal/commit/2071e36f3bbcf6fbe57f94a568f34812ad581c7e#diff-5f827d47769fbe9312e9e904c05fbc3eR53) along with the introduction of `metal::cascade`, but I can't remember why, most likely some compiler didn't want to cooperate.

I have made the following observation: Primary class templates with huge packs are expensive to instantiate. I can imagine that specialization (pattern matching `list<vals...>`) is more efficient.

> I don't know whether that change had any impact in benchmarks, but we could definitely attempt to revert it back to the previous implementation to avoid introducing `dcall`.

We can use `invoke` just fine.